logs_analyzer/app/scala/src/main/scala/com/databricks/apps/logs/ApacheAccessLog.scala

```diff
@@ -0,0 +1,42 @@
+package com.databricks.apps.logs
+
+/**
+  * An entry of Apache access log.
+  */
+
+case class ApacheAccessLog(ipAddress: String,
+                           clientIdentd: String,
+                           userId: String,
+                           dateTime: String,
+                           method: String,
+                           endpoint: String,
+                           protocol: String,
+                           responseCode: Int,
+                           contentSize: Long) extends Serializable {
+}
+
+object ApacheAccessLog {
+  private val PATTERN = """^(\S+) (\S+) (\S+) \[([\w:/]+\s[+\-]\d{4})\] "(\S+) (\S+) (\S+)" (\d{3}) (\d+)""".r
+
```

Comment:
Since Scaladoc is meant to be machine-readable, and only minimally human-readable, you could remove the blank lines so the comment is vertically compressed:

```
/** Parse log entry from a string.
  * @param log A string, typically a line from a log file
  * @return An entry of Apache access log, an empty value if cannot parse the line */
```